ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp

```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
+                 kernel,
+                 "\nkernel_strides=",
+                 kernel_strides,
+                 "\nkernel_dilations=",
+                 kernel_dilations,
+                 "\npads_begin=",
+                 pads_begin,
+                 "\npads_end=",
+                 pads_end);
+}
+
+template <typename T>
+struct Coord : public std::vector<T> {
+    Coord(const Shape& pads_begin) {
+        std::vector<T>::reserve(pads_begin.size());
+        for (const auto axis_padding : pads_begin) {
+            std::vector<T>::push_back(0 - axis_padding);
+        }
+    }
+
+    Coord(std::initializer_list<T>&& values) : std::vector<T>{std::move(values)} {}
+};
+
+bool elem_in_padding_area(const Coord<int>& kernel_position,
+                          const Coord<size_t>& kernel_offset,
+                          const Shape& data_shape) {
+    for (size_t dim = 0; dim < data_shape.size() - 2; ++dim) {
+        if (kernel_position[dim] + kernel_offset[dim] < 0 ||
+            kernel_position[dim] + kernel_offset[dim] >= data_shape[dim + 2]) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+template <typename T>
+Coord<T> next_kernel_position_2D(Coord<T> kernel_position,
+                                 const Shape& kernel,
+                                 const Strides& kernel_strides,
+                                 const Strides& kernel_dilations,
+                                 const Shape& data_shape,
+                                 const Shape& pads_begin,
+                                 const Shape& pads_end) {
+    // move the kernel horizontally one stride to the right
+    kernel_position[1] += kernel_strides[1];
+
+    // if the top-right corner of the kernel is outside of the padding area,
+    // move it back to the left and one stride down
+    if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+        kernel_position[1] = 0 - pads_begin[1];
+        kernel_position[0] += kernel_strides[0];
+    }
+
+    return kernel_position;
+}
+
+template <typename T>
+Coord<T> next_kernel_position_3D(Coord<T> kernel_position,
+                                 const Shape& kernel,
+                                 const Strides& kernel_strides,
+                                 const Strides& kernel_dilations,
+                                 const Shape& data_shape,
+                                 const Shape& pads_begin,
+                                 const Shape& pads_end) {
+    kernel_position[2] += kernel_strides[2];
+
+    if (kernel_position[2] + (kernel[2] - 1) * kernel_dilations[2] >= data_shape[4] + pads_end[2]) {
+        kernel_position[2] = 0 - pads_begin[2];
+        kernel_position[1] += kernel_strides[1];
+        if (kernel_position[1] + (kernel[1] - 1) * kernel_dilations[1] >= data_shape[3] + pads_end[1]) {
+            kernel_position[1] = 0 - pads_begin[1];
+            kernel_position[0] += kernel_strides[0];
+        }
```

Comment:
You can simply call `next_kernel_position_2D` instead

Reply:
If I call the 2D version I will get a 2D Coord object in return. I will have to merge it into the currently processed 3D version. This requires copying or moving the underlying integers which be less performant than having this slightly duplicated code.